Implement Constantinople fork

gurukamath opened the pull request:
(closes #434 )
(closes #435 )
(closes #436 )


Adding the Constantinople fork. The PR is still in the draft stage since it waits for the Byzantium Precomplies and will need to be patched on top of those. 

EIP-145, EIP-1052 and EIP-1234 are complete and can be reviewed. Some tests from EIP-1052 fail because they also need EIP-1014.


#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://upload.wikimedia.org/wikipedia/commons/d/d2/Cavy_eating_grass.jpg)

gurukamath commented:
@SamWilsn This is now ready for review.

I have marked the tests that depend on `create2` as slow for now. 

Also marked `stSStoreTest/InitCollision_d0g0v0.json` as slow. This is similar to the case that @voith pointed out some time back where a `current_target` has no nonce and balance but has some storage.

Bhargavasomu commented:
@gurukamath can you please give me permissions to your copy of the execution specs so as to add commits corresponding to `CREATE2`. Thanks!

voith commented on src/ethereum/constantinople/vm/interpreter.py:
```diff
@@ -164,6 +166,51 @@ def process_create_message(message: Message, env: Environment) -> Evm:
+    # It's expected that the creation operation works on empty storage. Hence
+    # we delete the storage and restore the account's state if there is an
+    # error in the initialization code execution.
+    destroy_storage(env.state, message.current_target)
```
For anyone reviewing this: The only difference between this method and `process_create_message` is this line. I was wondering if we could re-use `process_create_message` but I can't think of a way that would be specs appropriate.

voith commented on src/ethereum/constantinople/utils/address.py:
```diff
@@ -59,3 +59,32 @@ def compute_contract_address(address: Address, nonce: Uint) -> Address:
+    preimage = b"\xff" + address + salt + keccak256(call_data)
+    computed_address = keccak256(preimage)
+    canonical_address = computed_address[-20:]
+    padded_address = canonical_address.rjust(20, b"\x00")
```
Can you use [left_pad_zero_bytes](https://github.com/ethereum/execution-specs/blob/db01c78535df658e3a1e9eaa65a890cd501171a8/src/ethereum/utils/byte.py#L17-L33) utility method instead?

voith commented on src/ethereum/constantinople/vm/interpreter.py:
```diff
@@ -253,6 +300,11 @@ def execute_code(message: Message, env: Environment) -> Evm:
             PRE_COMPILED_CONTRACTS[evm.message.code_address](evm)
             return evm
+        else:
+            print(
```
Is this a leftover print statement from testing?

voith commented on src/ethereum/constantinople/vm/interpreter.py:
```diff
@@ -24,6 +24,8 @@
     account_has_code_or_nonce,
     begin_transaction,
     commit_transaction,
+    destroy_account,
```
I don't see this used anywhere.

voith commented on src/ethereum/constantinople/vm/interpreter.py:
```diff
@@ -164,6 +166,51 @@ def process_create_message(message: Message, env: Environment) -> Evm:
+        try:
+            evm.gas_left = subtract_gas(evm.gas_left, contract_code_gas)
+            ensure(len(contract_code) <= MAX_CODE_SIZE, OutOfGasError)
+        except OutOfGasError:
```
I think peter changed this exception to `ExceptionalHalt`

SamWilsn replied:
`OutOfGasError` is now a subclass of `ExceptionalHalt`: https://github.com/ethereum/execution-specs/blob/6a6f43417c73154d814262fe7dfb0f3f355d2248/src/ethereum/constantinople/vm/error.py#L52

voith commented on src/ethereum/constantinople/vm/interpreter.py:
```diff
@@ -164,6 +166,51 @@ def process_create_message(message: Message, env: Environment) -> Evm:
 
+def process_create2_message(message: Message, env: Environment) -> Evm:
+    """
+    Executes a call to create a smart contract via CREATE2 opcode.
```
Although I was the one who wrote "smart contract" in `process_create_message`, I'm not sure if it is appropriate for the specs. Can we say something like:
"Executes a call to deploy bytecode via CREATE2 opcode."?

voith commented on src/ethereum/constantinople/vm/instructions/system.py:
```diff
@@ -133,6 +139,107 @@ def create(evm: Evm) -> None:
+        push(evm.stack, U256(0))
+        evm.return_data = child_evm.output
+    else:
+        push(evm.stack, U256.from_be_bytes(child_evm.message.current_target))
```
I think you're missing the following line:
`evm.logs += child_evm.logs`

SamWilsn replied:
I believe you're correct @voith. Is this something we should add to #374 ?

SamWilsn commented:
I'm getting one failed test after I rebased and ported some stuff:

```
FAILED tests/constantinople/test_state_transition.py::test_create2[stCreate2/RevertInCreateInInitCreate2_d0g0v0.json
```

Did I break something?

gurukamath commented:
> I'm getting one failed test after I rebased and ported some stuff:
> 
> ```
> FAILED tests/constantinople/test_state_transition.py::test_create2[stCreate2/RevertInCreateInInitCreate2_d0g0v0.json
> ```
> 
> Did I break something?

@SamWilsn @voith This appears to me to be another one of those zero nonce, zero balance but having some storage accounts. Voith's [PR](https://github.com/ethereum/execution-specs/pull/466#issuecomment-1080834569) will fix this.

voith commented:
You're right. `RevertInCreateInInitCreate2_d0g0v0` is similar to `InitCollision_d0g0v0` and it should added to the incorrect upstream tests list.

gurukamath commented on src/ethereum/constantinople/vm/instructions/bitwise.py:
```diff
@@ -0,0 +1,232 @@
+
+def bitwise_shl(evm: Evm) -> None:
+    """
+    Shift right (SHL) operation of the top 2 elements of the stack. Pushes the
```
Included in commit

gurukamath commented on src/ethereum/constantinople/vm/instructions/bitwise.py:
```diff
@@ -0,0 +1,232 @@
+    value = pop(evm.stack)
+
+    evm.pc += 1
+    if shift >= 256:
+        push(evm.stack, U256(0))
+        return
+
+    shifted_value = (value << shift) % U256_CEIL_VALUE
+    push(evm.stack, U256(shifted_value))
```
Yes. This will indeed be simpler and more spec-like :) 
Included in commit

gurukamath commented on src/ethereum/constantinople/vm/instructions/bitwise.py:
```diff
@@ -0,0 +1,232 @@
+    value = pop(evm.stack)
+
+    evm.pc += 1
+    if shift >= 256:
+        push(evm.stack, U256(0))
+        return
+
+    shifted_value = value >> shift
+    push(evm.stack, shifted_value)
```
Yes. This will indeed be simpler and more spec-like :) 
Included in commit

gurukamath commented on src/ethereum/constantinople/vm/instructions/bitwise.py:
```diff
@@ -0,0 +1,232 @@
+    signed_value = value.to_signed()
+
+    evm.pc += 1
+    if shift >= 256:
+        if signed_value >= 0:
+            push(evm.stack, U256(0))
+            return
+        else:
+            push(evm.stack, U256(U256.MAX_VALUE))
+            return
+
+    shifted_value = signed_value >> shift
+    push(evm.stack, U256.from_signed(shifted_value))
```
Included in commit

gurukamath commented:
@SamWilsn 
Cleaned up the following parts of `tests/constantinople/test_state_transition.py`
- Unmarked all but 6 tests that were previously marked as slow, since create2 is now implemented. These 6 tests will be handled by #489 
- Removed the `test_create2` function to prevent create2 tests from running twice

@voith 
Found 4 more tests (in addition to the earlier 2) with zero nonce and balance but non-zero storage
`stSStoreTest/InitCollision_d0g0v0.json`
`stSStoreTest/InitCollision_d1g0v0.json`
`stSStoreTest/InitCollision_d2g0v0.json`
`stSStoreTest/InitCollision_d3g0v0.json`

Curiously however, `stSStoreTest/InitCollision_d2g0v0.json` passes for constantinople. Might be worth investigating.

SamWilsn commented:
I think this is good to merge now? :tada:

SamWilsn commented on src/ethereum/constantinople/state.py:
```diff
@@ -0,0 +1,499 @@
+
+def destroy_storage(state: State, address: Address) -> None:
+    """
+    Completely remove the account at `address` and all of its storage.
```
@Bhargavasomu

Bhargavasomu replied:
The main functionality here is to destroy the storage instead of destroying the account. I just introduced this function to insist that we are deleting the storage instead of the whole account. I guess the docstring here is wrong and will change this to be more meaningful of the function name.
